libappstream-builder/plugins/asb-plugin-sugar-activity.c

```diff
@@ -0,0 +1,286 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2016 Sam Parkinson <sam@sam.today>
+ *
+ * Licensed under the GNU Lesser General Public License Version 2.1
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+#include <string.h>
+#include <fnmatch.h>
+#include <gdk-pixbuf/gdk-pixbuf.h>
+
+#include <asb-plugin.h>
+
+#define __APPSTREAM_GLIB_PRIVATE_H
+#include <as-utils-private.h>
+#include <as-app-private.h>
+
+/**
+ * asb_plugin_get_name:
+ */
+const gchar *
+asb_plugin_get_name (void)
+{
+	return "sugar-activity";
+}
+
+#define _SUGAR_ACTIVITY_GLOB "/usr/share/sugar/activities/*/activity/activity.info"
+#define _SUGAR_ACTIVITY_ICON_GLOB "/usr/share/sugar/activities/*/activity/*"
+#define _SUGAR_ACTIVITY_LINFO_GLOB "/usr/share/sugar/activities/*/locale/*/activity.linfo"
+
+/**
+ * asb_plugin_add_globs:
+ */
+void
+asb_plugin_add_globs (AsbPlugin *plugin, GPtrArray *globs)
+{
+	// the _SUGAR_ACTIVITY_GLOB is a subset of the icon glob
+	asb_plugin_add_glob (globs, _SUGAR_ACTIVITY_ICON_GLOB);
+	asb_plugin_add_glob (globs, _SUGAR_ACTIVITY_LINFO_GLOB);
+}
+
+/**
+ * _asb_plugin_is_actinfo
+ */
+static gboolean
+_asb_plugin_is_actinfo (const gchar *filename)
+{
+	if (asb_plugin_match_glob (_SUGAR_ACTIVITY_GLOB, filename))
+		return TRUE;
+	return FALSE;
+}
+
+/**
+ * _asb_plugin_is_linfo
+ */
+static gboolean
+_asb_plugin_is_linfo (const gchar *filename)
+{
+	if (asb_plugin_match_glob (_SUGAR_ACTIVITY_LINFO_GLOB, filename))
+		return TRUE;
+	return FALSE;
+}
+
+/**
+ * asb_plugin_check_filename:
+ */
+gboolean
+asb_plugin_check_filename (AsbPlugin *plugin, const gchar *filename)
+{
+	return _asb_plugin_is_linfo (filename) ||
+	       _asb_plugin_is_actinfo (filename);
+}
+
+static gboolean
+_asb_plugin_load_sugar_icon (const gchar *path,
+			     AsIcon *ic) {
+	gsize len;
+	gchar *data = NULL;
+	GBytes *bytes = NULL;
+
+	if (!g_file_get_contents (path, &data, &len, NULL))
+		return FALSE;
+	printf ("ICON (got contents) = %s\n", path);
+
+	as_icon_set_kind (ic, AS_ICON_KIND_EMBEDDED);
+	bytes = g_bytes_new_take (data, len);
+	as_icon_set_data (ic, bytes);
+	as_icon_set_width (ic, 75);
```

Comment:
@samdroid-apps Haha, well:
From a metadata-purity point of view I disapprove that plugin's presence too. It's special-casing for a specific type of component and we really, really should not encourage that.

From a realistic perspective though, we have some metadata, and showing incomplete metadata is better than having no metadata. So I consider this plugin to be a good thing if the long-term plan is to get rid of it.
Now that I'm aware that this thing exists:
@hughsie what's the state of this feature? I see you added a "shell extension" component-type. Is this still an experimental feature? If not, it would make sense to add it too the spec, and I would also implement the metadata-extraction workaround in asgen, so Arch, Debian and Ubuntu can benefit from this.
That being said, "shell-extension" is a terrible name, because "shell" is waaay to generic. It's not a Plasma shell extension, not an extension to Bash or Zsh ;-)
Probably defining all of these things as component-type "addon" and have them extend the shell makes sense... (GS could special-case that).
Anyway, this isn't the best place to discuss this, I hope I'll think about moving it to the Mailinglist (otherwise -> IRC).